### Step 1 – Import Required Libraries

In [1]:
import pandas as pd
import numpy as np

### Step 2 – Load Dataset

In [2]:
df = pd.read_excel("Predictive modeling.xlsx")

In [3]:
# Create A Working Copy Of The Dataset

data = df.copy()

# Preview The First 5 Rows

data.head()

Payment                  ID            City gender  Age Nature_Sign  \
0  Not Paid  dc949d3e-db28-499f          Rajkot   Male   41       Eagle   
1  Not Paid  bb5b4f8a-6d18-47b5          Mumbai   Male   35      Turtle   
2  Not Paid  4058d768-2deb-4289       Bangalore   Male   39      Turtle   
3  Not Paid  52aa110e-4ff7-411b  Nadia district   Male   21      Turtle   
4  Not Paid  49a31680-6cf4-4a66           Delhi   Male   28      Turtle   

                 Industry          Profession  Credit_Score platform  ...  \
0    Education & Training           Professor         750.0  android  ...   
1                     NaN                 NaN         750.0      ios  ...   
2                     NaN                 NaN         864.0      ios  ...   
3                 Student   Trading+ business           NaN  android  ...   
4  Software & IT Services  Software Developer         753.0  android  ...   

  AltInv_Silver  Total_Liabilities  Housing_Loan  Auto_Loan  Personal_Loan  \
0           0.0                0.0           0.0        0.0            0.0   
1           0.0                0.0           0.0        0.0            0.0   
2           0.0          3100000.0     3100000.0        0.0            0.0   
3           0.0                NaN           NaN        NaN            NaN   
4           0.0          1310000.0           0.0        0.0      1310000.0   

   Consumer_Loan Credit_Card Education_Loan Gold_Loan Other_loan  
0            0.0         0.0            0.0       0.0        0.0  
1            0.0         0.0            0.0       0.0        0.0  
2            0.0         0.0            0.0       0.0        0.0  
3            NaN         NaN            NaN       NaN        NaN  
4            0.0         0.0            0.0       0.0        0.0  

[5 rows x 58 columns]

In [4]:
# Quick Overview

print(data.shape)       # Rows, Columns

(9610, 58)


In [5]:
print(data.info())      # Datatypes

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9610 entries, 0 to 9609
Data columns (total 58 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Payment                   9610 non-null   object 
 1   ID                        9610 non-null   object 
 2   City                      9610 non-null   object 
 3   gender                    9610 non-null   object 
 4   Age                       9610 non-null   int64  
 5   Nature_Sign               9610 non-null   object 
 6   Industry                  7310 non-null   object 
 7   Profession                7309 non-null   object 
 8   Credit_Score              8758 non-null   float64
 9   platform                  9443 non-null   object 
 10  device_model              9443 non-null   object 
 11  income_lac                9610 non-null   float64
 12  assest_cr                 9605 non-null   float64
 13  Liabilities_cr            9479 non-null   float64
 14  VUA     

In [6]:
data.describe().T  # Numerical Statistics

count          mean           std           min  \
Age                       9610.0  3.477086e+01  7.318480e+00  2.000000e+01   
Credit_Score              8758.0  7.731375e+02  5.936472e+01  3.000000e+02   
income_lac                9610.0  2.926415e+01  4.324730e+01  0.000000e+00   
assest_cr                 9605.0  1.585376e+00  8.598594e+00  0.000000e+00   
Liabilities_cr            9479.0  2.066425e-01  4.074527e-01  0.000000e+00   
VUA                       9605.0  1.789303e+00  8.646005e+00  1.000000e-07   
FBS_Latest                9577.0  5.297724e+01  1.315081e+01  1.100000e+01   
Total_Assets              9605.0  1.585376e+07  8.598594e+07  0.000000e+00   
Real_Estate               9605.0  9.020274e+06  5.390687e+07  0.000000e+00   
Equity                    9605.0  3.472819e+06  2.786137e+07  0.000000e+00   
Debt                      9605.0  2.672720e+06  1.017937e+07  0.000000e+00   
Gold                      9605.0  5.712959e+05  1.210067e+07  0.000000e+00   
Alt_Investments           9605.0  1.166466e+05  3.761234e+06  0.000000e+00   
non_Other                 9605.0  0.000000e+00  0.000000e+00  0.000000e+00   
RE_Residential            9605.0  7.411254e+06  1.955944e+07  0.000000e+00   
RE_Commercial             9605.0  1.609020e+06  4.159693e+07  0.000000e+00   
Equity_MF                 9605.0  1.362376e+06  4.319441e+06  0.000000e+00   
Equity_ESOPs              9605.0  3.867234e+05  9.449191e+06  0.000000e+00   
Equity_Stocks             9605.0  1.723719e+06  2.537990e+07  0.000000e+00   
Debt_EPF                  9605.0  8.069998e+05  2.247562e+06  0.000000e+00   
Debt_FD                   9605.0  5.948689e+05  5.749479e+06  0.000000e+00   
Debt_GovtSchemes          9605.0  5.997364e+04  4.072283e+05  0.000000e+00   
Debt_HybridMF             9605.0  5.517135e+04  8.496979e+05  0.000000e+00   
Debt_MF                   9605.0  1.043174e+05  1.173735e+06  0.000000e+00   
Debt_NPS                  9605.0  1.975960e+05  6.220328e+05  0.000000e+00   
Debt_Other                9605.0  6.936687e+04  1.131238e+06  0.000000e+00   
Debt_PPF                  9605.0  2.569750e+05  9.141735e+05  0.000000e+00   
Debt_Saving               9605.0  5.274509e+05  3.911624e+06  0.000000e+00   
Gold_Digital              9605.0  4.681334e+04  3.878088e+06  0.000000e+00   
Gold_Physical             9605.0  3.938409e+05  1.198013e+06  0.000000e+00   
Gold_SGB                  9605.0  1.306417e+05  8.167489e+06  0.000000e+00   
AltInv_Crypto             9605.0  6.760058e+04  3.742406e+06  0.000000e+00   
AltInv_P2P_OR_LoansGiven  9605.0  4.465312e+04  3.673851e+05  0.000000e+00   
AltInv_Silver             9605.0  4.392871e+03  4.894187e+04  0.000000e+00   
Total_Liabilities         9479.0  2.066425e+06  4.074527e+06  0.000000e+00   
Housing_Loan              9479.0  1.539127e+06  3.589976e+06  0.000000e+00   
Auto_Loan                 9479.0  1.411817e+05  4.573043e+05  0.000000e+00   
Personal_Loan             9479.0  2.119013e+05  1.227506e+06  0.000000e+00   
Consumer_Loan             9479.0  5.607219e+03  1.678287e+05  0.000000e+00   
Credit_Card               9479.0  1.102349e+04  8.067226e+04  0.000000e+00   
Education_Loan            9479.0  2.163032e+04  2.624372e+05  0.000000e+00   
Gold_Loan                 9479.0  5.752129e+03  8.519619e+04  0.000000e+00   
Other_loan                9479.0  1.302013e+05  1.067121e+06  0.000000e+00   

                                   25%           50%           75%  \
Age                       3.000000e+01  3.400000e+01  3.800000e+01   
Credit_Score              7.500000e+02  7.750000e+02  8.000000e+02   
income_lac                1.200000e+01  2.100000e+01  3.600000e+01   
assest_cr                 1.096954e-01  5.447709e-01  1.580000e+00   
Liabilities_cr            0.000000e+00  4.750000e-02  2.622500e-01   
VUA                       1.671533e-01  6.653000e-01  1.917000e+00   
FBS_Latest                4.400000e+01  5.400000e+01  6.300000e+01   
Total_Assets              1.096954

In [7]:
print(data.isnull().sum())  # Missing Values Check

Payment                        0
ID                             0
City                           0
gender                         0
Age                            0
Nature_Sign                    0
Industry                    2300
Profession                  2301
Credit_Score                 852
platform                     167
device_model                 167
income_lac                     0
assest_cr                      5
Liabilities_cr               131
VUA                            5
FBS_Latest                    33
Social_Generation              0
City_Group                     0
Income_Bucket                  0
Age_Bucket                     0
FBS_Bucket                     0
Cibil_Bucket                 852
Total_Assets                   5
Real_Estate                    5
Equity                         5
Debt                           5
Gold                           5
Alt_Investments                5
non_Other                      5
RE_Residential                 5
RE_Commerc

#### **Step - 3 Column Wise Cleaning Strategy**

##### Process Columns In Groups But Still Keep It Column-focused.

##### **Group 1 → Identifiers & Basic Info**
##### (ID, Payment, City, gender, Age, Nature_Sign)

##### **Group 2 → Work & Demographics**
##### (Industry, Profession, platform, device_model, Social_Generation, City_Group)

##### **Group 3 → Core Financials**
##### (income_lac, assest_cr, Liabilities_cr, VUA, FBS_Latest, Credit_Score, Cibil_Bucket)

##### **Group 4 → Buckets / Segmentation**
##### (Income_Bucket, Age_Bucket, FBS_Bucket)

##### **Group 5 → Assets**
##### (Total_Assets, Real_Estate, Equity, Debt, Gold, Alt_Investments, Equity_MF, Equity_ESOPs, Equity_Stocks, Debt_EPF, Debt_FD, Debt_GovtSchemes, Debt_HybridMF, Debt_MF, Debt_NPS, Debt_Other, Debt_PPF, Debt_Saving, Gold_Digital, Gold_Physical, Gold_SGB, AltInv_Crypto, AltInv_P2P_OR_LoansGiven, AltInv_Silver, non_Other)

##### **Group 6 → Liabilities / Loans**
##### (Total_Liabilities, Housing_Loan, Auto_Loan, Personal_Loan, Consumer_Loan, Credit_Card, Education_Loan, Gold_Loan, Other_loan)

### **Group 1 → Identifiers & Basic Info**
##### (ID, Payment, City, gender, Age, Nature_Sign)

#### **a) ID, Payment, City, gender, Age, Nature_Sign (Columns):-**

##### **Objective:-**

##### To Evaluate The Relevance Of Columns `ID, Payment, City, Gender, Age, And Nature_Sign` For The Analysis And Decide Which To Retain Or Drop.

##### Remove Redundant Or Non-Informative Fields `(ID, City, Nature_Sign)` To Simplify The Dataset And Prevent Accidental Use In Modeling.

##### Retain Key Demographic And Behavioral Features `(Age, Gender, Payment)` As They Provide Valuable Insights Into Customer Profiles And Payment Patterns.

##### **Observation:-**

##### The **ID** Column Is A Unique Identifier And Does Not Contribute To Predictive Modeling Or Analytics.
##### The **City** Column May Be Redundant Or Unnecessary For The Current Analysis, Especially If A Standardized **City_Group** Column Exists.
##### The **Nature_sign** Column May Contain Redundant Or Non-Informative Data For Predictive Modeling.
##### Dropping These Columns Helps Reduce Dataset Complexity And Avoids Accidental Misuse In Calculations.
##### The Columns **Age**, **Gender**, And **Payment** Contain Essential Information About Each Individual’s Demographic Profile And Payment Behavior & Columns Preserved Without Any Changes.

In [8]:
# Drop Unnecessary Columns
data = data.drop(columns=['ID', 'City','Nature_Sign'])

# Verify The Columns Have Been Dropped
print("Remaining Columns After Dropping ID, City And Nature_Sign:\n", data.columns.tolist())

Remaining Columns After Dropping ID, City And Nature_Sign:
 ['Payment', 'gender', 'Age', 'Industry', 'Profession', 'Credit_Score', 'platform', 'device_model', 'income_lac', 'assest_cr', 'Liabilities_cr', 'VUA', 'FBS_Latest', 'Social_Generation', 'City_Group', 'Income_Bucket', 'Age_Bucket', 'FBS_Bucket', 'Cibil_Bucket', 'Total_Assets', 'Real_Estate', 'Equity', 'Debt', 'Gold', 'Alt_Investments', 'non_Other', 'RE_Residential', 'RE_Commercial', 'Equity_MF', 'Equity_ESOPs', 'Equity_Stocks', 'Debt_EPF', 'Debt_FD', 'Debt_GovtSchemes', 'Debt_HybridMF', 'Debt_MF', 'Debt_NPS', 'Debt_Other', 'Debt_PPF', 'Debt_Saving', 'Gold_Digital', 'Gold_Physical', 'Gold_SGB', 'AltInv_Crypto', 'AltInv_P2P_OR_LoansGiven', 'AltInv_Silver', 'Total_Liabilities', 'Housing_Loan', 'Auto_Loan', 'Personal_Loan', 'Consumer_Loan', 'Credit_Card', 'Education_Loan', 'Gold_Loan', 'Other_loan']


### **Group 2 → Work & Demographics**
##### (Industry, Profession, platform, device_model, Social_Generation, City_Group)

#### **a) Industry Column**
##### **1. Objective**
##### The Industry Column Contains Information About The Sector Or Domain In Which Customers Are Employed Or Associated. It Is A Categorical Variable But Contains Duplicate Entries, Inconsistent Cases, Spelling Variations, And Special Characters. For Meaningful Analysis, We Need To Standardize And Categorize Industries Into Professional Groups.

##### **2. Observation**
##### **Case Sensitivity:** Some Values In Lowercase, Uppercase, Or Mixed (E.g., Software & IT Services, software and it services).
##### **Spelling Variations:** Minor Changes In Text (E.g., FMCG, fmcg, Fmcg).
##### **Duplicate Categories:** BFSI, NBFC, Banking Are Separate But Belong Under Financial Services.
##### **Special Characters & Spaces:** Extra Spaces (BFSI ) Or Symbols (Education & Training, education and training).
##### **Generic Terms:** Others, Not Disclosed, Student Which Need Tagging Or Separate Handling.

In [9]:
# View Sorted Unique Values

sorted(data['Industry'].dropna().unique())

[',Managing Consultant ',
 'AI & Data Science',
 'AIR INDIA',
 'AUTOMOBILE',
 'Accounting',
 'Advertising ',
 'Advertising & Marketing',
 'Advertising and Content Creation',
 'Aerospace',
 'Aerospace & Aviation',
 'Agriculture',
 'Agritech',
 'Airforce',
 'Airline',
 'Aluminium ',
 'Amazon',
 'Architecture',
 'Army',
 'Artist & Consultant',
 'Atlas Copco ',
 'Auto',
 'AutoMobile',
 'Automation',
 'Automobile',
 'Automobile ',
 'Automobile (EV)',
 'Automobile / Manufacturing',
 'Automobile Industry',
 'Automobiles',
 'Automotive',
 'Automotive ',
 'Automotive Industry',
 'Aviation',
 'BFSI',
 'BFSI ',
 'BPCL',
 'BPO',
 'BPO ',
 'BPO/KPO',
 'Banking',
 'Banking & Finance',
 'Businees',
 'Business',
 'Business & Tech Consultancy',
 'Business Processing Services',
 'Businessman',
 'Capital Asset',
 'Capital Goods',
 'Chartered Accountant',
 'Chemical',
 'Chemical ',
 'Chemical Industry ',
 'Chemical Manufacturing',
 'Chemicals',
 'Clothing',
 'Commercial',
 'Commercial Refrigeration ',
 'C

In [10]:
# Step a: Standardize Case And Strip Spaces

# Convert To Lowercase And Strip Spaces
data['Industry'] = data['Industry'].str.strip().str.lower()

print("Unique Industries (Raw, Lowercased):\n", sorted(data['Industry'].dropna().unique().T))

Unique Industries (Raw, Lowercased):
 [',managing consultant', 'accounting', 'advertising', 'advertising & marketing', 'advertising and content creation', 'aerospace', 'aerospace & aviation', 'agriculture', 'agritech', 'ai & data science', 'air india', 'airforce', 'airline', 'aluminium', 'amazon', 'architecture', 'army', 'artist & consultant', 'atlas copco', 'auto', 'automation', 'automobile', 'automobile (ev)', 'automobile / manufacturing', 'automobile industry', 'automobiles', 'automotive', 'automotive industry', 'aviation', 'banking', 'banking & finance', 'bfsi', 'bpcl', 'bpo', 'bpo/kpo', 'businees', 'business', 'business & tech consultancy', 'business processing services', 'businessman', 'capital asset', 'capital goods', 'chartered accountant', 'chemical', 'chemical industry', 'chemical manufacturing', 'chemicals', 'clothing', 'commercial', 'commercial refrigeration', 'commodities manufacturing', 'computer graphics', 'construction', 'construction equipment', 'consultancy', 'consult

In [11]:
# Step b: Banking & Financial Services (BFSI)

banking_map = {
    # BFSI Variations
    'bfsi': 'Banking & Financial Services',
    'bfsi ': 'Banking & Financial Services',

    # Banking & Finance
    'banking': 'Banking & Financial Services',
    'banking & finance': 'Banking & Financial Services',
    'dena bank': 'Banking & Financial Services',
    'Government Sector':'Banking & Financial Services',
    'government sector':'Banking & Financial Services',
    
    # Finance
    'finance': 'Banking & Financial Services',
    'finance and marketing': 'Banking & Financial Services',
    'financial consulting': 'Banking & Financial Services',
    'financial dealership': 'Banking & Financial Services',
    'financial services ': 'Banking & Financial Services',

    # Fintech
    'fintech': 'Banking & Financial Services',
    'fintech ': 'Banking & Financial Services',
    'marketing fintech': 'Banking & Financial Services',

    # NBFC
    'nbfc': 'Banking & Financial Services',
    'financial services': 'Banking & Financial Services',

    # Public & Private Sector (Under BFSI Umbrella)
    'public sector': 'Banking & Financial Services',
    'Public Sector': 'Banking & Financial Services',   # In Case Case-Sensitive Missed
    'PUBLIC SECTOR': 'Banking & Financial Services',
    'private sector': 'Banking & Financial Services',
    'psu': 'Banking & Financial Services',

    # Payment
    'payment services': 'Banking & Financial Services'
}

# Apply The Mapping

data['Industry'] = data['Industry'].replace(banking_map)

In [12]:
# Step c: Information Technology & Software Services:-

it_map = {
    # AI / Data
    'ai & data science': 'Information Technology & Software Services',
    'data generation': 'Information Technology & Software Services',
    'auto' : 'Information Technology & Software Services', 
    'automation' : 'Information Technology & Software Services',
    'process automation': 'Information Technology & Software Services',

    # BPO / KPO
    'bpo': 'Information Technology & Software Services',
    'bpo ': 'Information Technology & Software Services',
    'bpo/kpo': 'Information Technology & Software Services',
    'kpo': 'Information Technology & Software Services',
    'business processing services': 'Information Technology & Software Services',
    'Operations BPO':'Information Technology & Software Services',
    'OPS': 'Information Technology & Software Services',
    'operations bpo':'Information Technology & Software Services',
    'ops': 'Information Technology & Software Services',

    # Cybersecurity
    'cyber': 'Information Technology & Software Services',
    'cybersecurity ': 'Information Technology & Software Services',
    'cybersecurity' : 'Information Technology & Software Services',

    # Software / Technology
    'software': 'Information Technology & Software Services',
    'software & it services': 'Information Technology & Software Services',
    'technology': 'Information Technology & Software Services',
    'tech': 'Information Technology & Software Services',

    # IT Companies
    'tcs': 'Information Technology & Software Services',
    'PWC': 'Information Technology & Software Services',
    'delloite': 'Information Technology & Software Services',
    'amazon': 'Information Technology & Software Services', 
    'pwc' : 'Information Technology & Software Services', 
    'atlas copco' : 'Information Technology & Software Services',
    
    # Corporate / Services
    'corportate': 'Information Technology & Software Services',
    'service industry': 'Information Technology & Software Services',
    'service industry ': 'Information Technology & Software Services',
    'service sector': 'Information Technology & Software Services',
    'services': 'Information Technology & Software Services',
    'service provider to central govt on contractual basis': 'Information Technology & Software Services',

    # Misc Tech
    'computer graphics': 'Information Technology & Software Services',
    'process automation ': 'Information Technology & Software Services',
    'information services': 'Information Technology & Software Services',
}

# Apply IT Mapping

data['Industry'] = data['Industry'].replace(it_map)

In [13]:
# Step d: Automobile And Manufacturing Industries:-

auto_mfg_map = {
    # Automobile / Automotive
    'automobile': 'Automobile And Manufacturing Industries',
    'automobile ': 'Automobile And Manufacturing Industries',
    'automobile industry': 'Automobile And Manufacturing Industries',
    'automobiles': 'Automobile And Manufacturing Industries',
    'automotive': 'Automobile And Manufacturing Industries',
    'automotive ': 'Automobile And Manufacturing Industries',
    'automotive industry': 'Automobile And Manufacturing Industries',
    'automobile (ev)': 'Automobile And Manufacturing Industries',
    'automobile / manufacturing': 'Automobile And Manufacturing Industries',
    'ev': 'Automobile And Manufacturing Industries',
    'electric vehicle': 'Automobile And Manufacturing Industries',
    'tvs motors': 'Automobile And Manufacturing Industries',
    'recycling of automobile':'Automobile And Manufacturing Industries',
    'heavy machineries':'Automobile And Manufacturing Industries',

    # Manufacturing
    'manufacturing': 'Automobile And Manufacturing Industries',
    'manufacturing ': 'Automobile And Manufacturing Industries',
    'manufacturing industry': 'Automobile And Manufacturing Industries',
    'manufaturing': 'Automobile And Manufacturing Industries',
    'manufacturer': 'Automobile And Manufacturing Industries',
    'manufacturer of edible oil ': 'Automobile And Manufacturing Industries',
    'mfg': 'Automobile And Manufacturing Industries',
    'manufacturing industry': 'Automobile And Manufacturing Industries',
    'commodities manufacturing ': 'Automobile And Manufacturing Industries',
    'industrial tools': 'Automobile And Manufacturing Industries',
    'construction equipment': 'Automobile And Manufacturing Industries',
    'electronic manufacturing and service ': 'Automobile And Manufacturing Industries',
    'electrical ind': 'Automobile And Manufacturing Industries',
    'electrical mgf': 'Automobile And Manufacturing Industries',

    # Specific Sub-Sectors Under Manufacturing
    'aluminium ': 'Automobile And Manufacturing Industries',
    'steel industry': 'Automobile And Manufacturing Industries',
}

# Apply Mapping

data['Industry'] = data['Industry'].replace(auto_mfg_map)

In [14]:
# Step e: Automobile And Manufacturing Industries:-

edu_map = {
    # Core Education
    'education': 'Educational Sector & Training Institutes',
    'education ': 'Educational Sector & Training Institutes',
    'education & training': 'Educational Sector & Training Institutes',
    'educational sector ': 'Educational Sector & Training Institutes',
    'educational sector': 'Educational Sector & Training Institutes',
    'ed tech': 'Educational Sector & Training Institutes',
    'Learning & Education':'Educational Sector & Training Institutes',
    'learning & education':'Educational Sector & Training Institutes', 

    # Training / Mixed
    'soft skills training': 'Educational Sector & Training Institutes',
    'education and travel industry': 'Educational Sector & Training Institutes',
    'student':'Educational Sector & Training Institutes',
}

# Apply Mapping
data['Industry'] = data['Industry'].replace(edu_map)

In [15]:
# Step f: 'Management Consulting & Professional Business Services':-

consulting_map = {
    # Business Variants
    'businees': 'Management Consulting & Professional Business Services',
    'business': 'Management Consulting & Professional Business Services',
    'businessman': 'Management Consulting & Professional Business Services',
    'own business': 'Management Consulting & Professional Business Services',
    'family business': 'Management Consulting & Professional Business Services',
    'family Business': 'Management Consulting & Professional Business Services',
    'bUSINESS': 'Management Consulting & Professional Business Services',

    # Consulting Variants
    'consultancy': 'Management Consulting & Professional Business Services',
    'consultancy ': 'Management Consulting & Professional Business Services',
    'consultancy - chemicals': 'Management Consulting & Professional Business Services',
    'consultant': 'Management Consulting & Professional Business Services',
    'consultant ': 'Management Consulting & Professional Business Services',
    'consultation ': 'Management Consulting & Professional Business Services',
    'consulting': 'Management Consulting & Professional Business Services',
    'consulting ': 'Management Consulting & Professional Business Services',
    'management consultancy ': 'Management Consulting & Professional Business Services',
    'management consultancy' : 'Management Consulting & Professional Business Services',
    'management consultation': 'Management Consulting & Professional Business Services',
    'management consulting': 'Management Consulting & Professional Business Services',
    'strategy consultant': 'Management Consulting & Professional Business Services',
    'strategy consulting': 'Management Consulting & Professional Business Services',
    'strategy and consulting': 'Management Consulting & Professional Business Services',
    'research and consultancy': 'Management Consulting & Professional Business Services',
    'environment consulting': 'Management Consulting & Professional Business Services',
    'hr consulting ': 'Management Consulting & Professional Business Services',
    'sustainability management  ': 'Management Consulting & Professional Business Services',
    'waste management': 'Management Consulting & Professional Business Services',

    # Professional Services
    'professional services': 'Management Consulting & Professional Business Services',
    'artist & consultant': 'Management Consulting & Professional Business Services',
    'business & tech consultancy': 'Management Consulting & Professional Business Services',
    'managing consultant ': 'Management Consulting & Professional Business Services',
    ',managing consultant': 'Management Consulting & Professional Business Services',
}

# Apply mapping
data['Industry'] = data['Industry'].replace(consulting_map)

In [16]:
# Step g: Aerospace, Aviation & Railway Transportation:-

aerospace_aviation_railway_map = [
    'aerospace', 'aerospace & aviation', 'air india', 'airline', 'aviation',
    'railway', 'railway factory pvt ltd & partner at real estate business', 'railways'
]

# Apply mapping
data['Industry'] = data['Industry'].replace(aerospace_aviation_railway_map, 
                                            'Aerospace, Aviation & Railway Transportation')

In [17]:
# Step h: Entertainment, Media, Events & Digital Marketing:-

entertainment_map = {
    # Advertising & Marketing
    'advertising': 'Entertainment, Media, Events & Digital Marketing',
    'advertising & marketing': 'Entertainment, Media, Events & Digital Marketing',
    'advertising and content creation': 'Entertainment, Media, Events & Digital Marketing',
    'marketing': 'Entertainment, Media, Events & Digital Marketing',
    'marketing and sales': 'Entertainment, Media, Events & Digital Marketing',
    'market research': 'Entertainment, Media, Events & Digital Marketing',
    'media & marketing': 'Entertainment, Media, Events & Digital Marketing',

    # Design & Creative
    'desiging': 'Entertainment, Media, Events & Digital Marketing',
    'design': 'Entertainment, Media, Events & Digital Marketing',
    'design consultant': 'Entertainment, Media, Events & Digital Marketing',
    'designer': 'Entertainment, Media, Events & Digital Marketing',

    # Digital Marketing
    'digital marketing': 'Entertainment, Media, Events & Digital Marketing',
    'e-commerce digital marketing': 'Entertainment, Media, Events & Digital Marketing',

    # Entertainment & Events
    'entertainment': 'Entertainment, Media, Events & Digital Marketing',
    'event': 'Entertainment, Media, Events & Digital Marketing',
    'events': 'Entertainment, Media, Events & Digital Marketing',
    'film industry': 'Entertainment, Media, Events & Digital Marketing',
    'marriage lounge': 'Entertainment, Media, Events & Digital Marketing',
    'wedding photography': 'Entertainment, Media, Events & Digital Marketing',

    # Media & Communications
    'media & communications': 'Entertainment, Media, Events & Digital Marketing',
}

# Apply mapping
data['Industry'] = data['Industry'].replace(entertainment_map)

In [18]:
# Step i: Defense, Military & Armed Forces:-

defense_map = {
    'army': 'Defense, Military & Armed Forces',
    'indian army': 'Defense, Military & Armed Forces',
    'indian navy': 'Defense, Military & Armed Forces',
    'airforce': 'Defense, Military & Armed Forces',
    'defense': 'Defense, Military & Armed Forces',
    'defence': 'Defense, Military & Armed Forces',
    'defense and army': 'Defense, Military & Armed Forces',
    'defense and manufacturing': 'Defense, Military & Armed Forces',
    'merchant navy' : 'Defense, Military & Armed Forces'
}

# Apply Mapping
data['Industry'] = data['Industry'].replace(defense_map)

In [19]:
# Step j: E-Commerce Services & Engineering Services:-

ecom_merge_map = {
    'e- commerce': 'E-Commerce Services & Engineering Services',
    'e commerce': 'E-Commerce Services & Engineering Services',
    'ecommerce': 'E-Commerce Services & Engineering Services',
    'E-Commerce': 'E-Commerce Services & Engineering Services',
    'electric': 'E-Commerce Services & Engineering Services',
    'electrical': 'E-Commerce Services & Engineering Services',
    'e commerce services': 'E-Commerce Services & Engineering Services',
    'e commerece':'E-Commerce Services & Engineering Services',
    'e- commerce/ amazon':'E-Commerce Services & Engineering Services', 
    'e-com':'E-Commerce Services & Engineering Services', 
    'e-commerce':'E-Commerce Services & Engineering Services', 
    'multiple - e-commerce':'E-Commerce Services & Engineering Services',
    'internet / e- commerce': 'E-Commerce Services & Engineering Services',
    'electronic engineering services': 'E-Commerce Services & Engineering Services',
    'electronics engineering services': 'E-Commerce Services & Engineering Services',
    'multiple - e-commerce, education and travel industry': 'E-Commerce Services & Engineering Services',
}

# Apply Mapping
data['Industry'] = data['Industry'].replace(ecom_merge_map)

In [20]:
# Step k: Agriculture, Food & Consumer Goods (FMCG) Industry:-

agri_food_fmcg_map = {
    # Agriculture
    'agriculture': 'Agriculture, Food & Consumer Goods (FMCG) Industry',
    'agritech': 'Agriculture, Food & Consumer Goods (FMCG) Industry',

    # Consumer Goods / FMCG
    'consumer goods': 'Agriculture, Food & Consumer Goods (FMCG) Industry',
    'fmcg': 'Agriculture, Food & Consumer Goods (FMCG) Industry',
    'fmcg - captain fresh': 'Agriculture, Food & Consumer Goods (FMCG) Industry',
    'fmcg/edible oils': 'Agriculture, Food & Consumer Goods (FMCG) Industry',
    'fmcg\'': 'Agriculture, Food & Consumer Goods (FMCG) Industry',
    'tech, fmcg and trading': 'Agriculture, Food & Consumer Goods (FMCG) Industry',

    # Food & Beverages
    'food': 'Agriculture, Food & Consumer Goods (FMCG) Industry',
    'food and beverage': 'Agriculture, Food & Consumer Goods (FMCG) Industry',
    'food and beverages': 'Agriculture, Food & Consumer Goods (FMCG) Industry',
    'food distribution': 'Agriculture, Food & Consumer Goods (FMCG) Industry',
    'food industry': 'Agriculture, Food & Consumer Goods (FMCG) Industry',
    'edible oils' : 'Agriculture, Food & Consumer Goods (FMCG) Industry',
    'manufacturer of edible oil': 'Agriculture, Food & Consumer Goods (FMCG) Industry'
}

# Apply Mapping
data['Industry'] = data['Industry'].replace(agri_food_fmcg_map)

In [21]:
# Step l: Oil, Gas, Petrochemical & Power Energy Industry:-

oil_gas_power_map = {
    # Oil & Gas
    'bpcl': 'Oil, Gas, Petrochemical & Power Energy Industry',
    'ongc': 'Oil, Gas, Petrochemical & Power Energy Industry',
    'oil & gas': 'Oil, Gas, Petrochemical & Power Energy Industry',
    'oil & gas industry': 'Oil, Gas, Petrochemical & Power Energy Industry',
    'oil and gas': 'Oil, Gas, Petrochemical & Power Energy Industry',
    'oil industry': 'Oil, Gas, Petrochemical & Power Energy Industry',
    'oil refinery': 'Oil, Gas, Petrochemical & Power Energy Industry',

    # Petrochemicals
    'petro-chemical': 'Oil, Gas, Petrochemical & Power Energy Industry',
    'petrochemical': 'Oil, Gas, Petrochemical & Power Energy Industry',

    # Chemicals
    'chemical': 'Oil, Gas, Petrochemical & Power Energy Industry',
    'chemical industry': 'Oil, Gas, Petrochemical & Power Energy Industry',
    'chemical manufacturing': 'Oil, Gas, Petrochemical & Power Energy Industry',
    'chemicals': 'Oil, Gas, Petrochemical & Power Energy Industry',

    # Power & Energy
    'power': 'Oil, Gas, Petrochemical & Power Energy Industry',
    'power sector': 'Oil, Gas, Petrochemical & Power Energy Industry',
    'power generating': 'Oil, Gas, Petrochemical & Power Energy Industry',
    'power mfg': 'Oil, Gas, Petrochemical & Power Energy Industry',
    'energy': 'Oil, Gas, Petrochemical & Power Energy Industry',
    'energy sector': 'Oil, Gas, Petrochemical & Power Energy Industry',

    # Renewable Energy
    'renewable energy': 'Oil, Gas, Petrochemical & Power Energy Industry',
    'renewable energy sector': 'Oil, Gas, Petrochemical & Power Energy Industry',
    'renewable sector': 'Oil, Gas, Petrochemical & Power Energy Industry'
}

# Apply Mapping
data['Industry'] = data['Industry'].replace(oil_gas_power_map)

In [22]:
# Step m: Garments, Fashion, Footwear & Textile Industry:-

garments_fashion_textile_map = {
    # Clothing & Fashion
    'clothing': 'Garments, Fashion, Footwear & Textile Industry',
    'fashion': 'Garments, Fashion, Footwear & Textile Industry',
    'fashion & textile': 'Garments, Fashion, Footwear & Textile Industry',

    # Garments
    'garment business': 'Garments, Fashion, Footwear & Textile Industry',
    'garment industry': 'Garments, Fashion, Footwear & Textile Industry',

    # Footwear
    'footwear - bata india': 'Garments, Fashion, Footwear & Textile Industry',
    'sports / footwear': 'Garments, Fashion, Footwear & Textile Industry',

    # Textile
    'manufacturing and textile': 'Garments, Fashion, Footwear & Textile Industry',
    'textile': 'Garments, Fashion, Footwear & Textile Industry',
    'textiles': 'Garments, Fashion, Footwear & Textile Industry',
    'textile industry': 'Garments, Fashion, Footwear & Textile Industry',
    'textile manufacturing': 'Garments, Fashion, Footwear & Textile Industry',

    # Misc
    'material': 'Garments, Fashion, Footwear & Textile Industry',
    'sports': 'Garments, Fashion, Footwear & Textile Industry',
    'society manaagement': 'Garments, Fashion, Footwear & Textile Industry',
}

# Apply Mapping
data['Industry'] = data['Industry'].replace(garments_fashion_textile_map)

In [23]:
# Step n: Supply Chain, Logistics & Import-Export Transportation:-

supply_chain_logistics_map = {
    # Import & Export
    'export': 'Supply Chain, Logistics & Import-Export Transportation',
    'import & export': 'Supply Chain, Logistics & Import-Export Transportation',
    'import export of electronics': 'Supply Chain, Logistics & Import-Export Transportation',
    'import/export': 'Supply Chain, Logistics & Import-Export Transportation',
    'mobile access. - imports': 'Supply Chain, Logistics & Import-Export Transportation',

    # Logistics
    'logistics': 'Supply Chain, Logistics & Import-Export Transportation',
    'transportation & logistics': 'Supply Chain, Logistics & Import-Export Transportation',
    'logistics, construction, fmcg': 'Supply Chain, Logistics & Import-Export Transportation',

    # Supply Chain
    'supply chain': 'Supply Chain, Logistics & Import-Export Transportation',

    # Trading / Packaging / Product
    'trading': 'Supply Chain, Logistics & Import-Export Transportation',
    'product': 'Supply Chain, Logistics & Import-Export Transportation',
    'packaging': 'Supply Chain, Logistics & Import-Export Transportation',
    'shipping':  'Supply Chain, Logistics & Import-Export Transportation', 
    'shipping industry': 'Supply Chain, Logistics & Import-Export Transportation',
    'shiping' : 'Supply Chain, Logistics & Import-Export Transportation',
}

# Apply Mapping
data['Industry'] = data['Industry'].replace(supply_chain_logistics_map)

In [24]:
# Step o: Sports, Gaming & Retail Industry:-


sports_gaming_retail_map = {
    # Gaming & ESports
    'e sports': 'Sports, Gaming & Retail Industry',
    'gaming': 'Sports, Gaming & Retail Industry',
    'gaming and technology': 'Sports, Gaming & Retail Industry',
    'sports fantasy': 'Sports, Gaming & Retail Industry',

    # Retail
    'retail': 'Sports, Gaming & Retail Industry',
    'retail - e commerce': 'Sports, Gaming & Retail Industry',
    'retail entrepreneur': 'Sports, Gaming & Retail Industry',
    'retail industry': 'Sports, Gaming & Retail Industry',
    'retail sector': 'Sports, Gaming & Retail Industry',
    'retail, wholesale': 'Sports, Gaming & Retail Industry'
}

# Apply Mapping
data['Industry'] = data['Industry'].replace(sports_gaming_retail_map)

In [25]:
# Step p: Telecommunication, Hardware & Engineering Industry:-

telecom_hardware_engg_map = {
    # Engineering
    'electric engineering': 'Telecommunication, Hardware & Engineering Industry',
    'engineer': 'Telecommunication, Hardware & Engineering Industry',
    'engineering': 'Telecommunication, Hardware & Engineering Industry',
    'engineering consulting': 'Telecommunication, Hardware & Engineering Industry',
    'engineering products': 'Telecommunication, Hardware & Engineering Industry',
    'mechanical engineering': 'Telecommunication, Hardware & Engineering Industry',
    'mech': 'Telecommunication, Hardware & Engineering Industry',

    # Electronics & Hardware
    'electronic manufacturing and service': 'Telecommunication, Hardware & Engineering Industry',
    'electronics': 'Telecommunication, Hardware & Engineering Industry',
    'hardware': 'Telecommunication, Hardware & Engineering Industry',
    'hardware industry': 'Telecommunication, Hardware & Engineering Industry',
    'semi conductor designing': 'Telecommunication, Hardware & Engineering Industry',
    'semiconductor': 'Telecommunication, Hardware & Engineering Industry',

    # Telecommunication Variants
    'tele': 'Telecommunication, Hardware & Engineering Industry',
    'tele communications': 'Telecommunication, Hardware & Engineering Industry',
    'telecom': 'Telecommunication, Hardware & Engineering Industry',
    'telecom & ips': 'Telecommunication, Hardware & Engineering Industry',
    'telecom & isp': 'Telecommunication, Hardware & Engineering Industry',
    'telecom r&d': 'Telecommunication, Hardware & Engineering Industry',
    'telecome': 'Telecommunication, Hardware & Engineering Industry',
    'telecome and networking': 'Telecommunication, Hardware & Engineering Industry',
    'telecommunications': 'Telecommunication, Hardware & Engineering Industry'
}

# Apply Mapping
data['Industry'] = data['Industry'].replace(telecom_hardware_engg_map)

In [26]:
# Step q: Healthcare, Pharma & Medical Industry:-

healthcare_pharma_map = {
    # Core Healthcare & Pharma
    'medical': 'Healthcare, Pharma & Medical Industry',
    'health': 'Healthcare, Pharma & Medical Industry',
    'health care': 'Healthcare, Pharma & Medical Industry',
    'health care & pharma': 'Healthcare, Pharma & Medical Industry',
    'pharma': 'Healthcare, Pharma & Medical Industry',
    'medical devices': 'Healthcare, Pharma & Medical Industry',
    'doctor': 'Healthcare, Pharma & Medical Industry',

    # Wellness & Care
    'fitness': 'Healthcare, Pharma & Medical Industry',
    'nutrition and care': 'Healthcare, Pharma & Medical Industry',
    'petcare': 'Healthcare, Pharma & Medical Industry',

    # Research
    'research': 'Healthcare, Pharma & Medical Industry',
    'research and expert network': 'Healthcare, Pharma & Medical Industry',

    # Mixed category with healthcare
    'education,healthcare,construction': 'Healthcare, Pharma & Medical Industry',
    'travel & hospitality': 'Healthcare, Pharma & Medical Industry'
}

# Apply Mapping
data['Industry'] = data['Industry'].replace(healthcare_pharma_map)

In [27]:
# Step r: Legal, HR, Real Estate & Professional Services Industry:-

legal_hr_realestate_prof_map = {
    # Legal
    'law': 'Legal, HR, Real Estate & Professional Services Industry',
    'law business': 'Legal, HR, Real Estate & Professional Services Industry',
    'law firm': 'Legal, HR, Real Estate & Professional Services Industry',
    'lawyer': 'Legal, HR, Real Estate & Professional Services Industry',
    'legal': 'Legal, HR, Real Estate & Professional Services Industry',
    'legal department': 'Legal, HR, Real Estate & Professional Services Industry',
    'legal service': 'Legal, HR, Real Estate & Professional Services Industry',
    
    # HR & Recruitment
    'hr': 'Legal, HR, Real Estate & Professional Services Industry',
    'hr consulting': 'Legal, HR, Real Estate & Professional Services Industry',
    'human resource': 'Legal, HR, Real Estate & Professional Services Industry',
    'human resources': 'Legal, HR, Real Estate & Professional Services Industry',
    'recruitement': 'Legal, HR, Real Estate & Professional Services Industry',
    'recruitment': 'Legal, HR, Real Estate & Professional Services Industry',

    # Accounting, Trade & Merchant
    'accounting': 'Legal, HR, Real Estate & Professional Services Industry',
    'chartered accountant': 'Legal, HR, Real Estate & Professional Services Industry',
    'goods trading': 'Legal, HR, Real Estate & Professional Services Industry',
    'grain merchant': 'Legal, HR, Real Estate & Professional Services Industry',
    'plywood supply/trade': 'Legal, HR, Real Estate & Professional Services Industry',

    # Real Estate
    'real estate': 'Legal, HR, Real Estate & Professional Services Industry',
    'real estate advisory': 'Legal, HR, Real Estate & Professional Services Industry',

    # Creative & Personal Services
    'interior designer': 'Legal, HR, Real Estate & Professional Services Industry',
    'interior designing': 'Legal, HR, Real Estate & Professional Services Industry',
    'photography': 'Legal, HR, Real Estate & Professional Services Industry',
    'performing arts': 'Legal, HR, Real Estate & Professional Services Industry',
    'saloon': 'Legal, HR, Real Estate & Professional Services Industry',

    # Hospitality
    'hotel': 'Legal, HR, Real Estate & Professional Services Industry',
    'restraunt business': 'Legal, HR, Real Estate & Professional Services Industry',

    # Miscellaneous
    'house maker': 'Legal, HR, Real Estate & Professional Services Industry',
    'retired': 'Legal, HR, Real Estate & Professional Services Industry',
    'retired professional': 'Legal, HR, Real Estate & Professional Services Industry',
    'retirerd': 'Legal, HR, Real Estate & Professional Services Industry',
    'freelancing': 'Legal, HR, Real Estate & Professional Services Industry',
    'unemployed': 'Legal, HR, Real Estate & Professional Services Industry',
    'paytm salesman+ credit society advisor + own business of grocery': 'Legal, HR, Real Estate & Professional Services Industry',
    'security services': 'Legal, HR, Real Estate & Professional Services Industry',
    'architecture' : 'Legal, HR, Real Estate & Professional Services Industry',
}

# Apply Mapping
data['Industry'] = data['Industry'].replace(legal_hr_realestate_prof_map)

In [28]:
# Step s: Miscellaneous, Metals, Construction & Other Specialized Industries:-

miscellaneous_map = {
    # Metals & Mining
    'aluminium': 'Miscellaneous, Metals, Construction & Other Specialized Industries',
    'iron & steel': 'Miscellaneous, Metals, Construction & Other Specialized Industries',
    'metal & mining': 'Miscellaneous, Metals, Construction & Other Specialized Industries',
    'metals and mining': 'Miscellaneous, Metals, Construction & Other Specialized Industries',
    'mining': 'Miscellaneous, Metals, Construction & Other Specialized Industries',

    # Construction & Development
    'construction': 'Miscellaneous, Metals, Construction & Other Specialized Industries',
    'development': 'Miscellaneous, Metals, Construction & Other Specialized Industries',
    'home improvement': 'Miscellaneous, Metals, Construction & Other Specialized Industries',
    'modular furniture': 'Miscellaneous, Metals, Construction & Other Specialized Industries',
    'paint': 'Miscellaneous, Metals, Construction & Other Specialized Industries',
    'commercial refrigeration': 'Miscellaneous, Metals, Construction & Other Specialized Industries',

    # Capital & Assets
    'capital asset': 'Miscellaneous, Metals, Construction & Other Specialized Industries',
    'capital goods': 'Miscellaneous, Metals, Construction & Other Specialized Industries',
    'commodities manufacturing': 'Miscellaneous, Metals, Construction & Other Specialized Industries',

    # Insurance & Finance
    'insurance': 'Miscellaneous, Metals, Construction & Other Specialized Industries',
    'usa taxation': 'Miscellaneous, Metals, Construction & Other Specialized Industries',

    # NGO & International
    'ngo': 'Miscellaneous, Metals, Construction & Other Specialized Industries',
    'united nations works': 'Miscellaneous, Metals, Construction & Other Specialized Industries',
    'international political sciences': 'Miscellaneous, Metals, Construction & Other Specialized Industries',

    # Sustainability & Security
    'corporate security': 'Miscellaneous, Metals, Construction & Other Specialized Industries',
    'sustainability management': 'Miscellaneous, Metals, Construction & Other Specialized Industries',

    # Jewelry
    'jewelers shop': 'Miscellaneous, Metals, Construction & Other Specialized Industries',
    'jwellery': 'Miscellaneous, Metals, Construction & Other Specialized Industries',

    # Household Roles
    'homemaker': 'Miscellaneous, Metals, Construction & Other Specialized Industries',
    'housewife': 'Miscellaneous, Metals, Construction & Other Specialized Industries',

    # Professional / Other
    'consultation': 'Miscellaneous, Metals, Construction & Other Specialized Industries',
    'detailing': 'Miscellaneous, Metals, Construction & Other Specialized Industries',
    'hft': 'Miscellaneous, Metals, Construction & Other Specialized Industries',
    'immigration': 'Miscellaneous, Metals, Construction & Other Specialized Industries',
    'intellectual property rights': 'Miscellaneous, Metals, Construction & Other Specialized Industries',
    'localization': 'Miscellaneous, Metals, Construction & Other Specialized Industries',
    'offsore': 'Miscellaneous, Metals, Construction & Other Specialized Industries',
    'sabattical': 'Miscellaneous, Metals, Construction & Other Specialized Industries',
    'signage': 'Miscellaneous, Metals, Construction & Other Specialized Industries',
    'tyres': 'Miscellaneous, Metals, Construction & Other Specialized Industries',
    'water industry': 'Miscellaneous, Metals, Construction & Other Specialized Industries',
    'cvz': 'Miscellaneous, Metals, Construction & Other Specialized Industries',
    'na': 'Miscellaneous, Metals, Construction & Other Specialized Industries',
    'others': 'Miscellaneous, Metals, Construction & Other Specialized Industries',
    'commercial' : 'Miscellaneous, Metals, Construction & Other Specialized Industries',
}

# Apply Mapping
data['Industry'] = data['Industry'].replace(miscellaneous_map)

In [29]:
# Step t: Handling Missing Values In Industry Column
data['Industry'] = data['Industry'].fillna('Miscellaneous, Metals, Construction & Other Specialized Industries')

# Apply Mapping
print("Number Of Null Records After Filling:", data['Industry'].isnull().sum())

Number Of Null Records After Filling: 0


In [30]:
# Verification

print("Sample Values From Industry Column:\n", data['Industry'].head(10))

Sample Values From Industry Column:
 0             Educational Sector & Training Institutes
1    Miscellaneous, Metals, Construction & Other Sp...
2    Miscellaneous, Metals, Construction & Other Sp...
3             Educational Sector & Training Institutes
4           Information Technology & Software Services
5    Miscellaneous, Metals, Construction & Other Sp...
6                         Banking & Financial Services
7              Automobile And Manufacturing Industries
8    Miscellaneous, Metals, Construction & Other Sp...
9    Miscellaneous, Metals, Construction & Other Sp...
Name: Industry, dtype: object


#### **b) Profession (Columns):-**

##### **Objective**

##### **The Purpose Of This Step Is To:**

##### Identify And Correct Spelling Mistakes, Formatting Issues, And Duplicate Variations In The Profession Column.

##### Replace Null, Blank, Or Invalid Entries With "NaN".

##### Standardize Multiple Variations Of The Same Profession Into A Single Clean Category (E.g., Manager, Consultant, Engineer, Doctor).

##### Improve Data Quality For Further Analysis And Reporting.

##### **Observation**

##### Records Contained Inconsistencies Such As:

##### Extra Spaces → 'CEO' vs 'CEO '

##### Misspellings → 'Adminstration' vs 'Administration'

##### Abbreviations → 'Engg' vs 'Engineer'

##### Duplicates With Different Formats → 'Operation Manager', 'Operations manager ', 'OPERATIONS MANAGER'

##### Null / Blank / "NA" Entries

##### This Caused Inflated Unique Values And Made Grouping/analysis Unreliable.

In [31]:
# Unique Values Sorted In Ascending Order

unique_professions = sorted(data['Profession'].dropna().unique())
print("Unique Values In Profession Column (Sorted Ascending):\n", unique_professions)

Unique Values In Profession Column (Sorted Ascending):
 [' Tool Design Engineer', '.', '2D Supervisor', '2nd Officer ', '3D Animator', '3D Artist', '3D artist', 'AGM', 'AGM - Internal Audit Head ', 'AI Engineer', 'AI Researcher', 'AI development', 'ALM Director', 'ASIC Engineer', 'ASSISTANT MANAGER', 'AVP - Business Development', 'AVP - Infra Environment Manager ', 'AVP - Sales ', 'AVP: New Partnerships ', 'Able body seaman', 'Academic Alliance - Head ', 'Academic Consultant', 'Account Manager', 'Accountant', 'Accounting and Auditing', 'Accounting and Finance ', 'Accounting and finance ', 'Accounts & Finance Manager ', 'Acoustic', 'Acquisition Manager', 'Actor', 'Actuarial Consultant', 'Actuary', 'Admin', 'Admin ', 'Admin - Railways', 'Admin Head', 'Admin Professional', 'Admin manager', 'Administrative officer', 'Adminstration', 'Admission counselor', 'Admissions Counselor', 'Admistration', 'Advertising Manager', 'Advocate', 'Advocate ', 'Advocate, Business', 'Aerospace Engineer', 'Agr

In [32]:
# Count Of Null (NaN) Values

missing_count = data['Profession'].isnull().sum()
print("\nNumber Of Null Records In Profession Column:", missing_count)


Number Of Null Records In Profession Column: 2301


In [33]:
# Percentage Of Missing Values

total_rows = len(data)
print(f"Missing Percentage: {(missing_count/total_rows)*100:.2f}%")

Missing Percentage: 23.94%


In [34]:
# Replace Blank / NA-like alues with NaN
data['Profession'] = data['Profession'].replace(
    ['', '.', ' ', 'NA', 'Na', None],
    pd.NA
)

# Fill NaN with a default category
data['Profession'] = data['Profession'].fillna('Other Professions')

# Normalize text: lowercase & strip spaces
data['Profession'] = data['Profession'].str.strip().str.lower()

# Mapping Dictionary For Standardization
profession_map = {
    # CEO
    'ceo': 'CEO',
    'ceo ': 'CEO',

    # Administration
    'adminstration': 'Administration',
    'admistration': 'Administration',

    # Assistant Professor
    'assistant prof.': 'Assistant Professor',
    'assistant professor': 'Assistant Professor',
    'assosiate, associate': 'Associate',

    # Engineers
    'automation engg': 'Automation Engineer',
    'automation engineer': 'Automation Engineer',
    'civil engg': 'Civil Engineer',
    'civil engineer': 'Civil Engineer',
    'design engineer': 'Design Engineer',
    'design engineer ': 'Design Engineer',
    'design enggineer': 'Design Engineer',
    'electrical engg': 'Electrical Engineer',
    'electrical engineer': 'Electrical Engineer',
    'executive eng.': 'Executive Engineer',
    'executive engg': 'Executive Engineer',
    'eng': 'Engineer',
    'engg manager': 'Engineering Manager',
    'hardware engg': 'Hardware Engineer',
    'hardware engineer': 'Hardware Engineer',
    'lead engineer': 'Lead Engineer',
    'lead engineer ': 'Lead Engineer',
    'mechanical engineer': 'Mechanical Engineer',
    'mechanichal engineer': 'Mechanical Engineer',
    'software developer': 'Software Developer',
    'software developer ': 'Software Developer',
    'software eng': 'Software Engineer',
    'software engg': 'Software Engineer',
    'software engineer': 'Software Engineer',

    # Managerial Roles
    'branch manager - sales': 'Branch Sales Manager',
    'branch sales manager': 'Branch Sales Manager',
    'contract manager': 'Contracts Manager',
    'contracts manager ': 'Contracts Manager',
    'operation manager': 'Operations Manager',
    'operation manager': 'Operations Manager',
    'operations manager ': 'Operations Manager',
    'operational manager': 'Operations Manager',
    'operations manager': 'Operations Manager',
    'process manager': 'Process Manager',
    'process manager ': 'Process Manager',
    'product manager': 'Product Manager',
    'product manager ': 'Product Manager',
    'program manager': 'Program Manager',
    'program manager ': 'Program Manager',
    'project manager': 'Project Manager',
    'project manager ': 'Project Manager',
    'purchase manager': 'Purchasing Manager',
    'purchasing manager': 'Purchasing Manager',
    'regional manager': 'Regional Manager',
    'regional manager ': 'Regional Manager',
    'sales manager': 'Sales Manager',
    'sales manager ': 'Sales Manager',
    'marketing manager': 'Marketing Manager',
    'marketing manager ': 'Marketing Manager',
    'senior manager': 'Senior Manager',
    'senior manager ': 'Senior Manager',
    'senior manager - operations': 'Senior Manager - Operations',
    'senior manager - operations ': 'Senior Manager - Operations',
    'service manager': 'Service Manager',
    'service manager ': 'Service Manager',

    # Business
    'busines': 'Business',
    'business': 'Business',
    'business ': 'Business',
    'entrepreneur': 'Entrepreneur',
    'entrepreneur ': 'Entrepreneur',
    'entreprenuer': 'Entrepreneur',
    'entrenpreneurial': 'Entrepreneur',
    'entrenprenuer + sales': 'Entrepreneur',

    # Consultant
    'consultation': 'Consultant',
    'consulting': 'Consultant',
    'consutation': 'Consultant',
    'consultancy': 'Consultant',
    'consultant': 'Consultant',
    'consultant ': 'Consultant',
    'management consultant': 'Management Consultant',
    'management consulting': 'Management Consultant',
    'management consultant': 'Management Consultant',
    'strategy consultant': 'Strategy Consultant',
    'strategy consultation': 'Strategy Consultant',
    'it consultant': 'IT Consultant',
    'it consultant ': 'IT Consultant',

    # Clerical Roles
    'clerk': 'Clerk',
    'clerk ': 'Clerk',
    'clerical': 'Clerk',

    # Doctor
    'doctor': 'Doctor',
    'doctor ': 'Doctor',
    'medical officer': 'Medical Officer',
    'medical officer ': 'Medical Officer',
    'medical representative': 'Medical Sales Representative',
    'medical sales representative': 'Medical Sales Representative',

    # Analyst
    'data anaylst': 'Data Analyst',
    'data analyst': 'Data Analyst',
    'esg analyst': 'ESG Analyst',
    'esg analyst ': 'ESG Analyst',
    'quality anaylst': 'Quality Analyst',
    'quality analyst': 'Quality Analyst',
    'risk anaylst': 'Risk Analyst',
    'risk analyst': 'Risk Analyst',

    # HR & Human Resource
    'human resource': 'Human Resources',
    'human resources': 'Human Resources',
    'human resources ': 'Human Resources',
    'hr': 'Human Resources',
    'hr ': 'Human Resources',
    'hr consulting': 'HR Consultant',

    # Misc
    'executive assistant': 'Executive Assistant',
    'executive assistance': 'Executive Assistant',
    'freelance': 'Freelancer',
    'freelancer': 'Freelancer',
    'freelancer ': 'Freelancer',
    'freelancing': 'Freelancer',
    'freelencing': 'Freelancer',
    'lawyer': 'Lawyer',
    'lawyer ': 'Lawyer',
    'lawyer': 'Lawyer',
    'lawyer': 'Lawyer',
    'lawyer': 'Lawyer',
    'proff': 'Professor',
    'proffesor': 'Professor',
    'propreitor': 'Proprietor',
    'ux designer': 'UX Designer',
    'ux/ui designer': 'UX Designer',
    'ui/ux designer': 'UI/UX Designer',
    'ui/ux designer': 'UI/UX Designer',
    'assistant manager': 'Assistant Manager',
    'assistant manager ': 'Assistant Manager',
    'team lead': 'Team Lead',
    'team lead ': 'Team Lead',
    'team leader': 'Team Lead',
    'team lead': 'Team Lead',
}
# Apply Mapping
data['Profession'] = data['Profession'].replace(profession_map)

# Capitalize  
data['Profession'] = data['Profession'].apply(lambda x: x.title() if isinstance(x, str) else x)

# Check "Other Professions" Count
nan_count = (data['Profession'] == 'Other Professions').sum()
print(f"Count of 'Other Professions': {nan_count}")

# Preview
print(data['Profession'].unique()[:20])

Count of 'Other Professions': 2306
['Professor' 'Other Professions' 'Trading+ Business' 'Software Developer'
 'Credit Analyst' 'Process Engineer' 'Development Manager'
 'Electrical Engineer' 'Business Development' 'Financial Advisor'
 'It Consultant' 'Product Manager' 'Business Analyst' 'Architect'
 'Branch Operations Manager' 'Others' 'Cfo' 'Content Creation'
 'Sales Manager' 'Project Manager']


In [35]:
data['Profession'].isnull().sum()

0

#### **c) Platform (Column):-** 
##### **Objective**

##### To Clean And Standardize The Platform Column By Handling Missing Or Blank Values And Ensuring Consistency Across All Records For Reliable Downstream Analysis.

##### **Observations (Before Cleaning)**

##### Total Records: 9610

##### Unique Values (Raw): ['android', 'ios', '', NaN]

##### **Outcome (After Cleaning)**

##### Unique Values (Cleaned): ['android', 'ios', 'Other_OS']

##### Blank /NaN Records: Removed (Converted to 'Other_OS')

In [36]:
# Step a): Check Initial Unique Values
print("Unique Values Before Cleaning:\n", data['platform'].unique())

# Step b): Replace Blank ('') And NaN With 'Other_OS'
data['platform'] = data['platform'].replace('', np.nan)   
data['platform'] = data['platform'].fillna('Other_OS')         

# Step c): Verify Cleaning
print("Unique Values After Cleaning:\n", data['platform'].unique())
print("\nValue counts:\n", data['platform'].value_counts())

Unique Values Before Cleaning:
 ['android' 'ios' nan]
Unique Values After Cleaning:
 ['android' 'ios' 'Other_OS']

Value counts:
 platform
android     5693
ios         3750
Other_OS     167
Name: count, dtype: int64


In [37]:
data['platform'].isnull().sum()

0

#### **d) Device Model (Columns):-**

##### **Objective:**

##### To Clean And Standardize The Device_model Column In The Dataset By Accurately Extracting The Device Manufacturer (Brand) From Unstructured And Inconsistent Model Names. This Includes:

##### Handling Blank, Empty, Or NaN Records.

##### Correcting Case Inconsistencies.

##### Removing Special Characters And Unnecessary Text.

##### Mapping Similar Or Variant Brand Names To A Single Standardized Name (E.g., Mi, mi, xiaomi → Xiaomi).

##### **Observation:**

##### The Device_model Column Contains Multiple Patterns With Inconsistent Brand Name Representations.

##### Some Records Include The Brand Followed By :: And Model Details, While Others Are Blank Or NaN.

##### Brands Such As `Xiaomi, Apple, Honor, Huawei, Oppo, Oneplus, And Others` Appear In Multiple Formats.

##### Without Cleaning, It Is Difficult To Perform Brand-level Analysis Or Aggregation.

In [38]:
# Step a): Count Blanks And NaN
blank_count = (data['device_model'] == '').sum()
nan_count = data['device_model'].isna().sum()

print(f"Blank Records: {blank_count}")
print(f"NaN Records: {nan_count}")

# Step b): Get Unique Values Sorted In Ascending Order
sorted_device_models = sorted(data['device_model'].dropna().unique())
print("\nUnique Values (Sorted ascending):\n", sorted_device_models)

Blank Records: 0
NaN Records: 167

Unique Values (Sorted ascending):
 ['HONOR::ALI-NX1', 'HONOR::ANY-NX1', 'HONOR::COL-AL10', 'HONOR::COR-AL00', 'HONOR::ELI-NX9', 'HONOR::FNE-NX9', 'HONOR::REA-NX9', 'HUAWEI::INE-LX1r', 'HUAWEI::LYA-L29', 'HUAWEI::STK-L22', 'Infinix::Infinix X6517', 'Infinix::Infinix X6525B', 'Infinix::Infinix X663', 'Infinix::Infinix X663C', 'Infinix::Infinix X6711', 'Infinix::Infinix X671B', 'Infinix::Infinix X6731', 'Infinix::Infinix X6739', 'Infinix::Infinix X676B', 'Infinix::Infinix X682C', 'Infinix::Infinix X6832', 'Infinix::Infinix X6838', 'Infinix::Infinix X6851', 'Infinix::Infinix X6851B', 'Infinix::Infinix X6861', 'Infinix::Infinix X6871', 'Infinix::Infinix X689', 'Infinix::Infinix X693', 'Infinix::Infinix X695', 'Itel::itel P661N', 'LAVA::LAVA LXX503', 'LAVA::LAVA LXX504', 'LAVA::LAVA LXX505', 'LAVA::LAVA LXX506', 'LAVA::LAVA LXX507', 'Lenovo::Lenovo TB-8505X', 'Mi::M2012K11AI', 'Micromax::IN1', 'Micromax::IN_Note1', 'Nokia::Nokia 8.1', 'Nokia::Nokia G21', 'N

In [39]:
import re

brand_mapping = {
    "iphone": "Apple IPhone",
    "ipad": "Apple IPad",
    "asus": "Asus",
    "google": "Google Pixel",
    "honor": "Honor",
    "huawei": "Huawei",
    "infinix": "Infinix",
    "itel": "Itel",
    "lava": "Lava",
    "lenovo": "Lenovo",
    "lg": "LG",
    "micromax": "Micromax",
    "motorola": "Motorola",
    "nokia": "Nokia",
    "nothing": "Nothing",
    "oneplus": "OnePlus",
    "oppo": "Oppo",           
    "poco": "POCO",
    "realme": "Realme",
    "redmi": "Redmi",
    "samsung": "Samsung",
    "tecno": "TECNO",
    "vivo": "Vivo",
    "xiaomi": "Xiaomi",
    "iqoo": "iQOO",
}

def clean_device_model(model):
    """Standardize smartphone brand names from the device_model column."""
    if pd.isna(model) or str(model).strip() == "":
        return "NaN"

    text = str(model).strip()
    # Take The Part Before '::', Lowercase It
    prefix = text.split("::")[0].lower()

    for key, value in brand_mapping.items():
        if prefix.startswith(key):
            return value

    return text

# Apply
data["device_model"] = data["device_model"].apply(clean_device_model)

print(sorted(data["device_model"].unique()))

['Apple IPad', 'Apple IPhone', 'Asus', 'Google Pixel', 'Honor', 'Huawei', 'Infinix', 'Itel', 'LG', 'Lava', 'Lenovo', 'Mi::M2012K11AI', 'Micromax', 'Motorola', 'NaN', 'Nokia', 'Nothing', 'OnePlus', 'Oppo', 'POCO', 'Realme', 'Redmi', 'Samsung', 'TECNO', 'Vivo', 'Xiaomi', 'iQOO']


In [40]:
data["device_model"].isnull().sum()

0

#### **e) Social_Generation:-**
##### **Objective:-**

##### To Remove The Social_Generation Column, Which Contains Labels Such As Generation Z, Millennials, Generation X, And Baby Boomers, Since It Does Not Contribute Directly To Predictive Modeling And Can Be Derived Later If Needed.

##### **Observation**

##### The Column Consists Of Predefined Generational Categories. These Values Are Not Numeric, And They Don’t Add Significant Analytical Value For Our Current Objectives (E.g., Financial Asset Modeling).
##### Retaining This Feature Could Lead To Unnecessary Complexity.

In [41]:
# Step a) Drop Social_Generation Column

data.drop(columns=['Social_Generation'], inplace=True)

# Verify Removal
data.columns

Index(['Payment', 'gender', 'Age', 'Industry', 'Profession', 'Credit_Score',
       'platform', 'device_model', 'income_lac', 'assest_cr', 'Liabilities_cr',
       'VUA', 'FBS_Latest', 'City_Group', 'Income_Bucket', 'Age_Bucket',
       'FBS_Bucket', 'Cibil_Bucket', 'Total_Assets', 'Real_Estate', 'Equity',
       'Debt', 'Gold', 'Alt_Investments', 'non_Other', 'RE_Residential',
       'RE_Commercial', 'Equity_MF', 'Equity_ESOPs', 'Equity_Stocks',
       'Debt_EPF', 'Debt_FD', 'Debt_GovtSchemes', 'Debt_HybridMF', 'Debt_MF',
       'Debt_NPS', 'Debt_Other', 'Debt_PPF', 'Debt_Saving', 'Gold_Digital',
       'Gold_Physical', 'Gold_SGB', 'AltInv_Crypto',
       'AltInv_P2P_OR_LoansGiven', 'AltInv_Silver', 'Total_Liabilities',
       'Housing_Loan', 'Auto_Loan', 'Personal_Loan', 'Consumer_Loan',
       'Credit_Card', 'Education_Loan', 'Gold_Loan', 'Other_loan'],
      dtype='object')

#### **f) City_Group (Column);-**
##### **Objective**

##### To Retain The City_Group Column As Is Because It Represents Important Location Segmentation Such As `Mumbai, Delhi NCR, Hyderabad, Bengaluru, Tier II & Below, Pune, Ahmedabad, Chennai, Kolkata`.
##### These Values May Be Relevant For Customer Profiling, Market Segmentation, Or Trend Analysis.

##### **Observation**

##### The Column Contains Well-Defined City Clusters And No Noisy Or Unstructured Data.

##### There’s No Need For Transformation Or Cleaning At This Stage.

In [42]:
# Keep City_Group Column As Is

city_group_values = data['City_Group'].unique()

print("Unique Values In City_Group:", city_group_values)

Unique Values In City_Group: ['Tier II & below' 'Mumbai' 'Bengaluru' 'Delhi NCR' 'Pune' 'Kolkata'
 'Chennai' 'Hyderabad' 'Ahmedabad']


### Group 3 → Core Financials
##### (Income_lac, Assest_cr, Liabilities_cr, VUA, FBS_Latest, Credit_Score, Cibil_Bucket)

##### **Objective:-**

##### To Clean And Standardize The Core Financials Group Of Columns:
##### `assest_cr, income_lac, Liabilities_cr, VUA, FBS_Latest, Credit_Score, And Cibil_Bucket.`

##### All Numeric Columns Will Be Converted To Float With Proper Decimal Formatting.

##### `Cibil_Bucket (Categorical Column)` Will Have Missing Values Replaced With `"No_Credit_Score".`

##### **Observation:-**

##### `assest_cr, income_lac, Liabilities_cr, VUA, FBS_Latest, And Credit_Score` Hold Numerical Records In Decimal Format `(E.g., 0.74, 40.00, 45.65).`

##### Some Rows May Have `Blank, NaN, Or Null` Values That Must Be Replaced With `0.`

##### `Cibil_Bucket` Contains String Values: `Excellent, Good, Poor, And NA.`

##### `NA` In `Cibil_Bucket` Means No Credit History And Should Be Standardized As `No_Credit_Score.`

In [43]:
# List Of Numeric Columns To Clean

numeric_cols = ['assest_cr', 'income_lac', 'Liabilities_cr', 
                'VUA', 'FBS_Latest', 'Credit_Score']

# Convert To Numeric And Fill Blanks/NaN With 0

for col in numeric_cols:
    data[col] = pd.to_numeric(data[col], errors='coerce').fillna(0)

# Clean Categorical Column: Cibil_Bucket

data['Cibil_Bucket'] = (
    data['Cibil_Bucket']
    .replace(['NA', 'Na', 'na', None], 'No_Credit_Score')
    .fillna('No_Credit_Score')
)

# Display Cleaned Data

print("Sample Records After Cleaning Core Financial Columns:")
print(data[['assest_cr', 'income_lac', 'Liabilities_cr', 'VUA',
            'FBS_Latest', 'Credit_Score', 'Cibil_Bucket']].isnull().sum())

Sample Records After Cleaning Core Financial Columns:
assest_cr         0
income_lac        0
Liabilities_cr    0
VUA               0
FBS_Latest        0
Credit_Score      0
Cibil_Bucket      0
dtype: int64


### Group 4 → Buckets / Segmentation
##### (Income_Bucket, Age_Bucket, FBS_Bucket)

#### **a) Income_Bucket (Column):-**
 
##### **Objective:-**

##### To Retain The Income_Bucket Column Without Making Any Transformations, Since Its Values Are Already Standardized.

##### **Observation:-**

##### Income_Bucket Contains Predefined Ranges:
##### `['0L-10L', '10L-25L', '25L-50L', '>50L']`

##### There Are No Missing, Blank, Or Invalid Records.

##### No Conversion Or Mapping Is Required.

In [44]:
# Keep Income_bucket As It Is

sorted(data['Income_Bucket'].unique())

['0L-10L', '10L-25L', '25L-50L', '>50L']

#### **b) Age_Bucket (Columns):-**
##### **Objective:-**

##### Review Existing Age_Bucket Labels And Correct Unclear Or Incomplete Values (E.g., “45 Yrs & Above”).
##### Create Clear, Non Overlapping Buckets To Support Accurate Segmentation And Analysis.

##### **Observation:-**

##### Current Age_bucket Values Are Inconsistent, E.g.:

##### 24 Yrs And Below
##### 25 Yrs To 34 Yrs
##### 35 Yrs To 44 Yrs
##### 45 Yrs & Above (Ambiguous Upper Bound)

##### Missing Values May Also Exist In Some Records.

##### Need To Rename And Align Buckets With Clear Numerical Ranges.

##### **Proposed Standard Bins**
##### **(Corrected Label)**	 
##### 24 Yrs And Below	 
##### 25 Yrs To 34 Yrs	 
##### 35 Yrs To 44 Yrs	 
##### 45 Yrs To 60 Yrs	 
##### 60 Yrs & Above	 

In [45]:
def clean_age_bucket(val):
    """
    Standardize age groups in Age_Bucket.
    """
    if pd.isna(val) or str(val).strip() == "":
        return np.nan   # Keep NaN For Now If Needed

    text = str(val).strip().lower()

    if "24" in text and "below" in text:
        return "24 yrs and below"
    if "25" in text and "34" in text:
        return "25 yrs to 34 yrs"
    if "35" in text and "44" in text:
        return "35 yrs to 44 yrs"
    if text.startswith("45"):
        return "45 yrs to 60 yrs"
    if text.startswith("60"):
        return "60 yrs & Above"

    return val  # Fallback If Unexpected Value

# Apply The Function
data["Age_Bucket"] = data["Age_Bucket"].apply(clean_age_bucket)

print("Unique Age Buckets After Cleaning:")
print(sorted(data["Age_Bucket"].unique()))

Unique Age Buckets After Cleaning:
['24 yrs and below', '25 yrs to 34 yrs', '35 yrs to 44 yrs', '45 yrs to 60 yrs']


In [46]:
sorted(data['FBS_Bucket'].unique())

['0 To 50', '51 To 75', '76 To 100']

#### **c) FBS_Bucket (Column):-**
##### **Objective:-**

##### Eliminate The Fbs_bucket `(“FBS Score Bucket” or “Financial Behaviour Score Bucket.”)` Field Because Its Data Relates To Recovery/Collection Scores, Which Are Not Relevant To Our Current Analytics Or Predictive-modeling Work.
##### It’s A Categorical Field That Groups Customers Based On A Scoring System (E.g., 0 To 50, 51 To 75, 76 To 100).
##### This Type Of Score Is Usually Created By Banks/NBFCs To Measure Repayment Behaviour Or Collection Risk:

##### **Observation:-**

##### FBS_bucket Contains Values Such As:

  | Bucket      | Descriptions                                     |
  | ----------- | ------------------------------------------------ |
  | `0 To 50`   | Low Repayment Behaviour / High Collection Effort |
  | `51 To 75`  | Moderate Repayment Behaviour                     |
  | `76 To 100` | Good Repayment Behaviour / Low Collection Risk   |

##### These Ranges Measure Collection/Recovery Performance, Not Customer Demographics Or Financial Metrics.

##### Keeping This Column Adds Noise And Unnecessary Processing Overhead.

In [47]:
# Drop FBS_Bucket Column

data.drop(columns=["FBS_Bucket"], inplace=True)

### **Group 5 → Assets**
##### (Total_Assets, Real_Estate, Equity, Debt, Gold, Alt_Investments, Equity_MF, Equity_ESOPs, Equity_Stocks, Debt_EPF, Debt_FD, Debt_GovtSchemes, 
##### Debt_HybridMF, Debt_MF, Debt_NPS, Debt_Other, Debt_PPF, Debt_Saving, Gold_Digital, Gold_Physical, Gold_SGB, AltInv_Crypto, 
##### AltInv_P2P_OR_LoansGiven, AltInv_Silver, non_Other)

#### **a) Total Assets & Related Columns:-**
##### **Objective:-**

##### Review The Column `Total_Assets` Along With Its Component Columns
##### (`Real_Estate, Equity, Debt, Gold, Alt_Investments`).

##### Verify That The Sum Of The Individual Asset Columns Equals The Value In `Total_Assets.`

##### Standardize The Numeric Format Of All These Columns (E.g., 444.000) For Easier Calculations And Reporting.

##### Drop `Total_Assets` If It Is Only A Derived Total (No Additional Information Beyond The Sum Of Its Components).

##### **Observation:-**

##### Total_Assets Contains Large Whole-number Values Such As 345678, 556777789, 1567890, 6752345678.

##### The Sum Of `Real_Estate, Equity, Debt, Gold, And Alt_Investments` Is Equal To (Or Very Close To) The Value In `Total_Assets.`

##### Keeping Both `Total_Assets` And Its Component Columns Would Create Redundancy.

##### Converting All Figures To A Decimal Format `(345.000)` Makes The Dataset Easier To Read, Validate, And Aggregate.

##### **Proposed Action:-**

##### Convert The Numeric Data In `Total_Assets, Real_Estate, Equity, Debt, Gold, Alt_Investments` Into A Standard Float Format With Three Decimal Places.

##### Remove The `Total_Assets` Column Since Its Information Can Be Reproduced From The Remaining Columns.

In [48]:
# Asset Columns

asset_cols = ['Real_Estate', 'Equity', 'Debt', 'Gold', 'Alt_Investments']

# Divide By 1,000,000 To Express In Millions

data[asset_cols] = data[asset_cols].apply(pd.to_numeric, errors='coerce') / 1_000_000

# Round To 3 Decimals

data[asset_cols] = data[asset_cols].round(3)

# Force Display With 3 Decimals

pd.set_option('display.float_format', '{:.3f}'.format)

print(data[['Real_Estate', 'Equity', 'Debt', 'Gold', 'Alt_Investments']].head(10))

   Real_Estate  Equity  Debt  Gold  Alt_Investments
0        5.000   0.350 2.007 0.000            0.000
1        0.000   4.700 2.300 0.000            0.012
2       16.000   0.891 2.795 0.051            0.000
3        0.000   1.802 0.300 0.000            0.000
4        0.000   0.585 1.400 0.000            0.000
5        0.000   2.650 0.980 0.000            0.000
6        0.132   1.233 4.972 0.200            0.000
7        0.000   0.400 1.784 0.300            0.000
8        0.000   4.000 1.570 0.500            0.000
9        4.000   0.000 0.272 0.000            0.000


In [49]:
# Drop Total_Assets If It's Only A Sum Of Others

data.drop(columns=['Total_Assets'], inplace=True)

#### **b) Non_Other (Column):-**

##### Remove Empty / Unused Column – Non_Other

##### **Objective**

##### Inspect The Column Non_Other For Data Availability.

##### If The Column Has 0 Non-Null Values (Completely Empty), Drop It To Keep The Dataset Clean.

##### **Observation**

##### Data.info() Shows Non_Other Exists But Contains 0 Non-null Values.

##### Keeping An Empty Column Serves No Purpose And Adds Clutter.

In [50]:
data.drop(columns=['non_Other'], inplace=True)

#### **c) Standardizing & Aggregating Investment (Columns):-**
##### [`RE_Residential, RE_Commercial, Equity_MF, Equity_ESOPs, Equity_Stocks, Debt_EPF, Debt_FD, Debt_GovtSchemes, Debt_HybridMF, Debt_MF, Debt_NPS, Debt_Other, Debt_PPF, Debt_Saving, Gold_Digital, Gold_Physical, Gold_SGB, AltInv_Crypto, AltInv_P2P_OR_LoansGiven, AltInv_Silver`]

##### **Objective:-**

##### Consolidate Granular Asset Fields (E.g., Equity_MF, Debt_FD, Gold_Physical, Etc.) Into Five Master Columns:

##### `-> Real_Estate`

##### `-> Equity`

##### `-> Debt`

##### `-> Gold`

##### `-> Alt_Investments`

##### Replace Blank / Null Values With 0 For Consistency.

##### Convert Values Into Millions With Three Decimals To Simplify Analysis And Reporting.

##### Remove Redundant Sub-Columns After Aggregation.

##### **Observation:-**

##### Many Asset Categories Were Split Across Multiple Columns, Creating Redundancy And Complicating Analysis.

##### Several Records Contained Missing Or Blank Values, Which Could Distort Totals If Not Handled.

##### **After Aggregation:-**

##### All Sub-Columns Successfully Merged Into Their Respective Category Columns.

##### Values Are Now Standardized In Million-Decimal Format (E.g., 45.000 = 45M).

##### Simplified Structure Makes It Easier To Compare Asset Classes And Validate Against Total_assets.

##### The Dataset Is Now More Compact And Ready For Modeling, Dashboards, Or Reporting.

In [51]:
# a) Defined Groups:-

# Real Estate, Equity,Debt, Gold, Alternative Investments

real_estate_cols = ['RE_Residential', 'RE_Commercial']
equity_cols = ['Equity_MF', 'Equity_ESOPs', 'Equity_Stocks']
debt_cols = ['Debt_EPF','Debt_FD','Debt_GovtSchemes','Debt_HybridMF',
             'Debt_MF','Debt_NPS','Debt_Other','Debt_PPF','Debt_Saving']
gold_cols = ['Gold_Digital','Gold_Physical','Gold_SGB']
alt_cols = ['AltInv_Crypto','AltInv_P2P_OR_LoansGiven','AltInv_Silver']

# b) Convert To Numeric & Replace Blanks With 0:-

data[real_estate_cols + equity_cols + debt_cols + gold_cols + alt_cols] = (
    data[real_estate_cols + equity_cols + debt_cols + gold_cols + alt_cols]
      .apply(pd.to_numeric, errors='coerce')
      .fillna(0)
)

In [52]:
# c) Create One Column Per Category (Sum Row Wise)

data["Real_Estate"]   = data[real_estate_cols].sum(axis=1)
data["Equity"]        = data[equity_cols].sum(axis=1)
data["Debt"]          = data[debt_cols].sum(axis=1)
data["Gold"]          = data[gold_cols].sum(axis=1)
data["Alt_Investments"] = data[alt_cols].sum(axis=1)

# d) Convert To Millions With 3 Decimals

cols_final = ["Real_Estate","Equity","Debt","Gold","Alt_Investments"]

data[cols_final] = (data[cols_final] / 1_000_000).round(3)
pd.set_option('display.float_format', '{:.3f}'.format)
print(data[cols_final].head())

   Real_Estate  Equity  Debt  Gold  Alt_Investments
0        5.000   0.350 2.007 0.000            0.000
1        0.000   4.700 2.300 0.000            0.012
2       16.000   0.891 2.795 0.051            0.000
3        0.000   1.802 0.300 0.000            0.000
4        0.000   0.585 1.400 0.000            0.000


In [53]:
# e) Drop The Original Granular Columns:-

data.drop(columns=real_estate_cols + equity_cols + debt_cols + gold_cols + alt_cols, inplace=True)

### **Group 6 → Liabilities / Loans**
##### (Total_Liabilities, Housing_Loan, Auto_Loan, Personal_Loan, Consumer_Loan, Credit_Card, Education_Loan, Gold_Loan, Other_loan)

#### **a) Liabilities & Loan Data – Consolidation And Cleaning:-**

##### **Objective:-**

##### Review All Columns Related To Customer Liabilities And Loans.

##### Remove Redundant Column Total_liabilities (Its Value Equals The SUM Of Individual Loan Types).

##### Replace Any NaN / Blank Values With 0 For Accurate Numeric Analysis.

##### Merge Sub-Categories That Represent The Same Type Of Loan Into A Single Column (E.g., Any Variants Of “Personal Loan”).

##### Keep Final Figures In Millions (M), Rounded To 3 Decimals, To Maintain Consistency With Assets.

##### **Columns Considered**

| Group               | Columns                          |
| ------------------- | -------------------------------- |
| Master (To Drop)    | `Total_Liabilities`              |
| Housing             | `Housing_Loan`                   |
| Vehicle             | `Auto_Loan`                      |
| Personal / Consumer | `Personal_Loan`, `Consumer_Loan` |
| Revolving Credit    | `Credit_Card`                    |
| Education           | `Education_Loan`                 |
| Gold-Backed         | `Gold_Loan`                      |
| Others              | `Other_loan`                     |


##### Note:- Any Extra Aliases (E.g., “Consumer_Loan” = “Personal_loan”) Will Be Merged Into The Parent Bucket.

##### **Observation:-**

##### Total_Liabilities Is Just An Aggregation Of All Other Loan Columns → Safe To Drop.

##### Some Records Contain NaN, Blanks Or None; These Need To Be Filled With 0.

##### A Few Categories Overlap:

##### Personal_loan And Consumer_loan → Merged As Personal_loan.

##### No Duplicate For Housing, Auto, Credit Card, Education, Gold, Other.

##### **After Cleaning:-**

##### Every Loan Column Will Hold Its Value In M (E.g., 1250000 → 1.250 M).

##### A Single Row Total = Housing + Auto + Personal + Creditcard + Education + Gold + Other.

In [54]:
# Cleaning & Merging Loans 

loan_cols = [
    "Total_Liabilities", "Housing_Loan", "Auto_Loan",
    "Personal_Loan", "Consumer_Loan", "Credit_Card",
    "Education_Loan", "Gold_Loan", "Other_loan"
]

# Fill Missing / Blank With 0

data[loan_cols] = data[loan_cols].replace(["", " ", "NaN", np.nan], 0).astype(float)

# Merge Consumer_Loan → Personal_Loan

data["Personal_Loan"] = data["Personal_Loan"] + data["Consumer_Loan"]

# Convert To Millions (3 Decimals)

for col in ["Housing_Loan", "Auto_Loan", "Personal_Loan",
            "Credit_Card", "Education_Loan", "Gold_Loan", "Other_loan"]:
    data[col] = data[col] / 1_000_000
    data[col] = data[col].round(3)

# Print Final DataFrame 

print("\n Cleaned Loan Data (in Millions):")
print(sorted(data[["Housing_Loan","Auto_Loan","Personal_Loan",
          "Credit_Card","Education_Loan","Gold_Loan","Other_loan"]].head()))


 Cleaned Loan Data (in Millions):
['Auto_Loan', 'Credit_Card', 'Education_Loan', 'Gold_Loan', 'Housing_Loan', 'Other_loan', 'Personal_Loan']


In [55]:
# Drop Total_Liabilities & Consumer_Loan

data.drop(columns=["Total_Liabilities", "Consumer_Loan"], inplace=True)

In [56]:
# Checking Missing Values In Each Column

data.isnull().sum()

Payment            0
gender             0
Age                0
Industry           0
Profession         0
Credit_Score       0
platform           0
device_model       0
income_lac         0
assest_cr          0
Liabilities_cr     0
VUA                0
FBS_Latest         0
City_Group         0
Income_Bucket      0
Age_Bucket         0
Cibil_Bucket       0
Real_Estate        0
Equity             0
Debt               0
Gold               0
Alt_Investments    0
Housing_Loan       0
Auto_Loan          0
Personal_Loan      0
Credit_Card        0
Education_Loan     0
Gold_Loan          0
Other_loan         0
dtype: int64

##### Save The Final Cleaned Dataset As A CSV File

In [57]:
# Save The Final Cleaned Dataframe As A CSV File

data.to_csv("Cleaned_Predictive_Modeling_Data.csv", index=False)
print("File Saved Successfully As 'Cleaned_Predictive_Modeling_Data.csv'")

File Saved Successfully As 'Cleaned_Predictive_Modeling_Data.csv'
